<a href="https://colab.research.google.com/github/sp2005-im/Google-Solution-Challenge/blob/main/Bioinformatics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#Sri Rama Jayam
!pip install chembl_webresource_client

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.2/55.2 kB 2.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.4/61.4 kB 5.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.4/66.4 kB 5.3 MB/s eta 0:00:00


In [2]:
#imports
import pandas as pd
from chembl_webresource_client.new_client import new_client

In [3]:
#Search for Target Protein
#Target search for coronavirus
target = new_client.target
target_query = target.search('coronavirus')
targets = pd.DataFrame.from_dict(target_query)
targets

,cross_references,organism,pref_name,score,species_group_flag,target_chembl_id,target_components,target_type,tax_id
0,[],Coronavirus,Coronavirus,17.0,False,CHEMBL613732,[],ORGANISM,11119
1,[],Feline coronavirus,Feline coronavirus,14.0,False,CHEMBL612744,[],ORGANISM,12663
2,[],Murine coronavirus,Murine coronavirus,14.0,False,CHEMBL5209664,[],ORGANISM,694005
3,[],Canine coronavirus,Canine coronavirus,14.0,False,CHEMBL5291668,[],ORGANISM,11153
4,[],Human coronavirus 229E,Human coronavirus 229E,13.0,False,CHEMBL613837,[],ORGANISM,11137
5,[],Human coronavirus OC43,Human coronavirus OC43,13.0,False,CHEMBL5209665,[],ORGANISM,31631
6,[],Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,10.0,False,CHEMBL3927,"[{'accession': 'P0C6U8', 'component_descriptio...",SINGLE PROTEIN,694009
7,[],Middle East respiratory syndrome-related coron...,Middle East respiratory syndrome-related coron...,9.0,False,CHEMBL4296578,[],ORGANISM,1335626
8,[],Severe acute respiratory syndrome-related coro...,Replicase polyprotein 1ab,4.0,False,CHEMBL5118,"[{'accession': 'P0C6X7', 'component_descriptio...",SINGLE PROTEIN,694009
9,[],Severe acute respiratory syndrome coronavirus 2,Replicase polyprotein 1ab,4.0,False,CHEMBL4523582,"[{'accession': 'P0DTD1', 'component_descriptio...",SINGLE PROTEIN,2697049


In [4]:
#select and retrieve biological activity for SARS coronavirus 3C-like proteinase
selected_target = targets.target_chembl_id[6]
selected_target

'CHEMBL3927'

In [5]:
activity = new_client.activity
res = activity.filter(target_chembl_id = selected_target).filter(standard_type="IC50")

In [6]:
df = pd.DataFrame.from_dict(res)

In [7]:
df.head(5)

,action_type,activity_comment,activity_id,activity_properties,assay_chembl_id,assay_description,assay_type,assay_variant_accession,assay_variant_mutation,bao_endpoint,...,target_organism,target_pref_name,target_tax_id,text_value,toid,type,units,uo_units,upper_value,value
0,None,None,1480935,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,7.2
1,None,None,1480936,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,9.4
2,None,None,1481061,[],CHEMBL830868,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,13.5
3,None,None,1481065,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,13.11
4,None,None,1481066,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,2.0


In [8]:
df.standard_type.unique()

array(['IC50'], dtype=object)

In [9]:
#Saving the bioactivity data to a CSV file -> bioactivity_data.csv
df.to_csv('bioactivity_data.csv', index = "False") #- We don't want the resulting index number in the csv file

In [10]:
from google.colab import drive
drive.mount('/content/gdrive/', force_remount = True)

Mounted at /content/gdrive/


In [11]:
!mkdir "/content/gdrive/My Drive/Colab Notebooks/data"

mkdir: cannot create directory ‘/content/gdrive/My Drive/Colab Notebooks/data’: File exists


In [12]:
!cp bioactivity_data.csv "/content/gdrive/My Drive/Colab Notebooks/data"

In [13]:
!ls -l "/content/gdrive/My Drive/Colab Notebooks/data"

total 137
-rw------- 1 root root 139682 Feb  6 09:12 bioactivity_data.csv


In [14]:
!ls

bioactivity_data.csv  gdrive  sample_data


In [15]:
!head -n 5 bioactivity_data.csv

,action_type,activity_comment,activity_id,activity_properties,assay_chembl_id,assay_description,assay_type,assay_variant_accession,assay_variant_mutation,bao_endpoint,bao_format,bao_label,canonical_smiles,data_validity_comment,data_validity_description,document_chembl_id,document_journal,document_year,ligand_efficiency,molecule_chembl_id,molecule_pref_name,parent_molecule_chembl_id,pchembl_value,potential_duplicate,qudt_units,record_id,relation,src_id,standard_flag,standard_relation,standard_text_value,standard_type,standard_units,standard_upper_value,standard_value,target_chembl_id,target_organism,target_pref_name,target_tax_id,text_value,toid,type,units,uo_units,upper_value,value
0,,,1480935,[],CHEMBL829584,In vitro inhibitory concentration against SARS coronavirus main protease (SARS CoV 3C-like protease),B,,,BAO_0000190,BAO_0000357,single protein format,Cc1noc(C)c1CN1C(=O)C(=O)c2cc(C#N)ccc21,,,CHEMBL1139624,Bioorg Med Chem Lett,2005,"{'bei': '18.28', 'le': '0.33', 'lle': '3.25', 's

In [29]:
l = df.standard_value
p = 0
for i in l:
  if p == None:
    p+=1
print(p)

0


In [22]:
df2 = df[df.standard_value.notna()]
df2.head(5)

,action_type,activity_comment,activity_id,activity_properties,assay_chembl_id,assay_description,assay_type,assay_variant_accession,assay_variant_mutation,bao_endpoint,...,target_organism,target_pref_name,target_tax_id,text_value,toid,type,units,uo_units,upper_value,value
0,None,None,1480935,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,7.2
1,None,None,1480936,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,9.4
2,None,None,1481061,[],CHEMBL830868,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,13.5
3,None,None,1481065,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,13.11
4,None,None,1481066,[],CHEMBL829584,In vitro inhibitory concentration against SARS...,B,None,None,BAO_0000190,...,Severe acute respiratory syndrome-related coro...,SARS coronavirus 3C-like proteinase,694009,None,None,IC50,uM,UO_0000065,None,2.0


In [34]:
l1 = df2.standard_value
print(len(l1))
t = 0
for i in l1:
  if i == None:
    t+=1
print(t)

245
0


In [30]:
#labelling compounds as active, inactive or intermediate based on IC 50 value
bioactivity_class = []
for i in df2.standard_value:
  if float(i)>=10000:
    bioactivity_class.append("inactive")
  elif float(i)<=1000:
    bioactivity_class.append("active")
  else:
    bioactivity_class.append("intermediate")


In [35]:
print(df2.molecule_chembl_id) #It won;t be 247 because we removed the null values

0       CHEMBL187579
1       CHEMBL188487
2       CHEMBL185698
3       CHEMBL426082
4       CHEMBL187717
           ...      
242    CHEMBL4590273
243    CHEMBL5436771
244    CHEMBL2365410
245    CHEMBL5436771
246    CHEMBL2365410
Name: molecule_chembl_id, Length: 245, dtype: object


**The above result shows 245 entries not 247 entries as two None values were removed**

In [36]:
mol_cid = []
for i in df2.molecule_chembl_id:
  mol_cid.append(i)

In [38]:
len(mol_cid)

245

In [44]:
df2.canonical_smiles

,canonical_smiles
0,Cc1noc(C)c1CN1C(=O)C(=O)c2cc(C#N)ccc21
1,O=C1C(=O)N(Cc2ccc(F)cc2Cl)c2ccc(I)cc21
2,O=C1C(=O)N(CC2COc3ccccc3O2)c2ccc(I)cc21
3,O=C1C(=O)N(Cc2cc3ccccc3s2)c2ccccc21
4,O=C1C(=O)N(Cc2cc3ccccc3s2)c2c1cccc2[N+](=O)[O-]
...,...
242,Cc1cccc2nc(CSC(=S)NCc3cccnc3)cn12
243,S=C([S-])NCc1cccnc1.[K+]
244,CC(C)C[C@H](NC(=O)OCc1ccccc1)C(=O)N[C@@H](CC1C...
245,S=C([S-])NCc1cccnc1.[K+]


In [45]:
#Cannonical Smiles - Simplified Molecular Input Line Entry System (SMILES) - Although unique for each structure, dependent on the cannonicalization
# Algorithm used
canonical_smiles = []
for i in df2.canonical_smiles:
  canonical_smiles.append(i)

In [47]:
len(canonical_smiles)

245

In [57]:
standard_value = []
for i in df2.standard_value:
  standard_value.append(i)
len(standard_value)

245

In [58]:
data_tuples = list(zip(mol_cid, canonical_smiles, bioactivity_class, standard_value))
df3 = pd.DataFrame(data_tuples, columns = ['molecule_chembl_id', 'canonical_smiles', 'bioactivity_class', 'standard_value'])

In [59]:
df3

,molecule_chembl_id,canonical_smiles,bioactivity_class,standard_value
0,CHEMBL187579,Cc1noc(C)c1CN1C(=O)C(=O)c2cc(C#N)ccc21,intermediate,7200.0
1,CHEMBL188487,O=C1C(=O)N(Cc2ccc(F)cc2Cl)c2ccc(I)cc21,intermediate,9400.0
2,CHEMBL185698,O=C1C(=O)N(CC2COc3ccccc3O2)c2ccc(I)cc21,inactive,13500.0
3,CHEMBL426082,O=C1C(=O)N(Cc2cc3ccccc3s2)c2ccccc21,inactive,13110.0
4,CHEMBL187717,O=C1C(=O)N(Cc2cc3ccccc3s2)c2c1cccc2[N+](=O)[O-],intermediate,2000.0
...,...,...,...,...
240,CHEMBL4590273,Cc1cccc2nc(CSC(=S)NCc3cccnc3)cn12,active,380.19
241,CHEMBL5436771,S=C([S-])NCc1cccnc1.[K+],active,165.0
242,CHEMBL2365410,CC(C)C[C@H](NC(=O)OCc1ccccc1)C(=O)N[C@@H](CC1C...,active,161.0
243,CHEMBL5436771,S=C([S-])NCc1cccnc1.[K+],active,165.96


In [60]:
df3.to_csv('bioactivity_preprocessed_data.csv', index = False)

In [61]:
!ls -l

total 172
-rw-r--r-- 1 root root 139682 Feb  6 09:07 bioactivity_data.csv
-rw-r--r-- 1 root root  20699 Feb  6 11:09 bioactivity_preprocessed_data.csv
drwx------ 6 root root   4096 Feb  6 09:07 gdrive
drwxr-xr-x 1 root root   4096 Feb  4 14:22 sample_data


In [62]:
!cp 'bioactivity_preprocessed_data.csv' "/content/gdrive/My Drive/Colab Notebooks/data"

In [63]:
!ls

bioactivity_data.csv  bioactivity_preprocessed_data.csv  gdrive  sample_data


In [64]:
!ls "/content/gdrive/My Drive/Colab Notebooks/data"

bioactivity_data.csv  bioactivity_preprocessed_data.csv
